validation: detect witness stripping without re-running Script checks

darosior opened the pull request:
Since it was introduced in 4eb515574e1012bc8ea5dafc3042dcdf4c766f26 (#18044), the detection of a stripped witness relies on running the Script checks 3 times. In the worst case, this consists in running Script validation for every single input 3 times.

Detection of a stripped witness is necessary because in this case wtxid==txid, and the transaction's wtxid must not be added to the reject filter or it could allow a malicious peer to interfere with txid-based orphan resolution as used in 1p1c package relay.

However it is not necessary to run Script validation to detect a stripped witness (much less so doing it 3 times in a row). There are 3 types of witness program: defined program types (Taproot, P2WPKH and P2WSH), undefined types, and the Pay-to-anchor carve-out.

For defined program types, Script validation with an empty witness will always fail (by consensus). For undefined program types, Script validation is always going to fail regardless of the witness (by standardness). For P2A, an empty witness is never going to lead to a failure.

Therefore it holds that we can always detect a stripped witness without re-running Script validation. However this might lead to more "false positives" (cases where we return witness stripping for an otherwise invalid transaction) than the existing implementation. For instance a transaction with one P2PKH input with an invalid signature and one P2WPKH input with its witness stripped. The existing implementation would treat it as consensus invalid while the implementation in this PR would always consider it witness stripped.

h/t AJ: this essentially implements a variant of https://github.com/bitcoin/bitcoin/pull/33066#issuecomment-3135258539.

darosior committed:
validation: detect witness stripping without re-running Script checks

Since it was introduced in 4eb515574e1012bc8ea5dafc3042dcdf4c766f26 (#18044), the detection of a
stripped witness relies on running the Script checks 3 times. In the worst case, this consists in
running Script validation 3 times for every single input.

Detection of a stripped witness is necessary because in this case wtxid==txid, and the transaction's
wtxid must not be added to the reject filter or it could allow a malicious peer to interfere with
txid-based orphan resolution as used in 1p1c package relay.

However it is not necessary to run Script validation to detect a stripped witness (much less so
doing it 3 times in a row). There are 3 types of witness program: defined program types (Taproot,
P2WPKH, P2WSH), undefined types, and the Pay-to-anchor carve-out.

For defined program types, Script validation with an empty witness will always fail (by consensus).
For undefined program types, Script validation is always going to fail regardless of the witness (by
standardness). For P2A, an empty witness is never going to lead to a failure.

Therefore it holds that we can always detect a stripped witness without re-running Script validation.
However this might lead to more "false positives" (cases where we return witness stripping for an
otherwise invalid transaction) than the existing implementation. For instance a transaction with one
P2PKH input with an invalid signature and one P2WPKH input with its witness stripped. The existing
implementation would treat it as consensus invalid while the implementation in this commit would
always consider it witness stripped.

ajtowns reviewed (commented):
Approach ACK

darosior commented on src/validation.cpp:
```diff
@@ -902,8 +902,24 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
+        if (m_pool.m_opts.require_standard && !IsWitnessStandard(tx, m_view)) {
+            return state.Invalid(TxValidationResult::TX_WITNESS_MUTATED, "bad-witness-nonstandard");
+        }
+    } else {
```
I considered it and decided against because 1) whether to add the txid to the reject filter seems unrelated to standardness (and surprising that `-acceptnonstdtxn=1` would now break 1p1c in the adversarial case) and 2) a failure here means the witness is invalid by consensus, #29843 would already not allow it today on master.

darosior commented on src/validation.cpp:
```diff
@@ -902,8 +902,24 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
+        std::vector<uint8_t> program;
+        for (const auto& txin: tx.vin) {
+            const auto& prev_spk{m_view.AccessCoin(txin.prevout).out.scriptPubKey};
+            if (prev_spk.IsWitnessProgram(version, program) && !prev_spk.IsPayToAnchor(version, program)) {
```
I think this would be a bit confusing because it would need to return `true` for undefined program types, whereas we don't know yes if this type will require a witness or not (presumably yes, but hey). I also think it's brief enough to not warrant its own function. Happy to change it, but it seems cleaner this way.

ajtowns replied:
I think that's fine -- if `require_standard` is true, then undefined program types will have already given an INPUTS_NOT_STANDARD response; if it's not true and #29843 has been merged, then we'll accept any witness including a null witness for undefined program types, so `RequiresNonEmptyWitness` should return false for them. That can change to true when they're defined in a soft-fork, but that's soft-fork behaviour, so should be fine.

But worrying about it in 29843 rather than here seems best.

darosior commented on src/validation.cpp:
```diff
@@ -902,8 +902,24 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     }
```
self nit: if i retouch, 0e22a1401c7edee8946f404dd0deb59a94231340's message is missing a word at the end of the second sentence.

darosior replied:
Done.

nervana21 commented on src/validation.cpp:
```diff
@@ -902,8 +902,24 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     }
 
     // Check for non-standard witnesses.
```
Small nit. This comment could be updated now that we're checking for both `nonstandard` and `stripped` witnesses

```suggestion
    // Check for witness-related policy violations.
```

glozow reviewed (commented):
approach ACK - I'm not super confident in my script knowledge but I can't think of why you couldn't detect witness stripped this way.

darosior commented:
Good catch.

darosior committed:
policy: introduce a helper to detect whether a transaction spends Segwit outputs

We will use this helper in later commits to detect witness stripping without having
to execute every input Script three times in a row.

darosior committed:
qa: test witness stripping in p2p_segwit

A stripped witness is detected as a special case in mempool acceptance to make sure we do not add
the wtxid (which is =txid since witness is stripped) to the reject filter. This is because it may
interfere with 1p1c parent relay which currently uses orphan reconciliation (and originally it was
until wtxid-relay was widely adopted on the network.

This commit adds a test for this special case in the p2p_segwit function test, both when spending
a native segwit output and when spending a p2sh-wrapped segwit output.

Thanks to Eugene Siegel for pointing out the p2sh-wrapped detection did not have test coverage by
finding a bug in a related patch of mine.

glozow commented on src/policy/policy.cpp:
```diff
@@ -337,6 +337,43 @@ bool IsWitnessStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+            std::vector <std::vector<uint8_t>> stack;
+            if (!EvalScript(stack, txin.scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker{}, SigVersion::BASE)
+                || stack.empty()) {
+                return false;
```
Oh I see. I would prefer returning witness-stripped in that scenario. It feels like this function returns a not-quite-accurate result in order to accommodate a different function that must run after it to catch a different error... a bit brittle.

ajtowns replied:
Shouldn't this mimic `VerifyScript` more closely? ie:

 * check the scriptSig is push-only
 * check the redeemScript matches the hash in the p2sh prev_spk
 * check the scriptSig doesn't push other things on the stack (ie `stack.size() == 1`)

All those are permanent errors for the txid (fixing them requires changing the scriptSig which changes the txid), so the rejection would be valid to cache.

> Currently we check scripts sequentially from first input to last. If input N is consensus-invalid and input N+1 is witness stripped we will return a consensus-invalid error.

I don't think that's the right idea? We call:

 1. `CheckInputScripts(tx, scriptVerifyFlags)`
 2. if that fails, and we have no witness, we check `CheckInputScripts(tx, scriptVerifyFlags - WITNESS,CLEANSTACK)` succeeds
 3. if it does, we check `CheckInputScripts(tx, scriptVerifyFlags - CLEANSTACK) fails
 4. if so, we return `WITNESS_STRIPPED`

But each of these consider every input; so consensus failures outside of witness validation will fail at step 2, so we'll return the error from step 1.

Because this PR returns witness stripped results earlier, we should expect to get different results than the current code; eg, for a tx without witness data, where the first input is a p2wpkh spend and the second input is a p2sh script with a bad redeemScript, currently I think we'll see a script-verification-error due to the first input, while after this patch I'd expect to see a witness stripped error.

darosior replied:
Hmm and if we do so, we don't even need to carve out an exception for P2A, because using P2A with an empty witness can never result in a failure.

I'm going to implement that now.

ajtowns replied:
The logic when we see "WITNESS_STRIPPED" is "this peer sent us garbage, that we have to ignore" -- to me, it's better to spend as little CPU as possible on garbage, so that means we should detect WITNESS_STRIPPED as soon and as cheaply as possible, then just ignore the tx. If we could have drawn some useful conclusions from the garbage (eg "INPUTS_NOT_STANDARD" or any other policy/consensus failure that different witness data couldn't fix), that's fine, but probably not at the cost of doing much extra processing -- caching the error will only help if we had another misbehaving peer wanting to send the same invalid tx.

> In fact i think your point hints that we should instead check for this after every other check.

So I think my point is the opposite -- we should check for witness stripped early, and not bother very much with doing the other checks even if they're easy. Just identify that the node's giving us garbage asap and move on.

Crypt-iQ replied:
> So I think my point is the opposite -- we should check for witness stripped early, and not bother very much with doing the other checks even if they're easy.

I agree.

I double-checked and the current code does not change behavior.

The early check in an earlier iteration of this PR changed behavior in two ways:
- consensus-invalid for non-witness reasons could be converted to witness-stripped, but this does not matter as they are consensus-invalid and a malicious peer can't interfere with honest relay.
- a tx that is TX_RECONSIDERABLE that a peer relayed without the witness was reported as TX_RECONSIDERABLE. With the early check, the tx is instead (correctly imo) reported as witness-stripped. I also think this does not matter in practice. See [comment](https://github.com/bitcoin/bitcoin/pull/33066#discussion_r2251819585).

Crypt-iQ commented on src/policy/policy.cpp:
```diff
@@ -337,6 +337,43 @@ bool IsWitnessStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+                || stack.empty()) {
+                return false;
+            }
+            // P2SH-wrapped Taproot programs are invalid by consensus. P2SH-wrapped Anchor program is non-standard.
```
I think P2SH-wrapped taproot programs are non-standard according to BIP341 text (says they remain unencumbered) and [here](https://github.com/bitcoin/bitcoin/blob/d1b583181dcc31613cd586d63329cb4c4a586972/src/script/interpreter.cpp#L1949)?

darosior commented:
I changed the call site for `SpendsNonAnchorWitnessProg` to be where we detect witness stripping today. Besides being a smaller diff this minimizes the potential for behaviour changes. Note this PR still slightly changes behaviour (see last commit's message and https://github.com/bitcoin/bitcoin/pull/33105#discussion_r2254632919).

Crypt-iQ commented:
Replying out of thread to not derail the thread conversation.

> This would make sure we don't weaken any existing standardness or consensus error into a witness-stripped one.

In current master, these are the consensus errors (I think?) that can be returned as witness-stripped:
- SCRIPT_ERR_WITNESS_MALLEATED (changes txid)
- SCRIPT_ERR_WITNESS_MALLEATED_P2SH (changes txid)
- SCRIPT_ERR_WITNESS_PROGRAM_MISMATCH (for P2WPKH, not the case for P2WSH)
- SCRIPT_ERR_WITNESS_PROGRAM_WRONG_LENGTH

The other errors require a witness and would fail the `CheckInputScripts` logic early. I might be misunderstanding the comment / being a bit pedantic and maybe you're referring to not adding even more errors that can be converted into witness-stripped?

darosior commented:
> maybe you're referring to not adding even more errors that can be converted into witness-stripped?

Yes this is what i meant. There are cases today where we would return WITNESS_STRIPPED for an otherwise invalid transaction (such as it could only be made valid by changing its txid, and we are missing the opportunity of adding its txid to the reject filter). This PR adds more such cases. This is minimized in the latest iteration (https://github.com/bitcoin/bitcoin/pull/33105#issuecomment-3156334619) by leaving the witness detection at the same site where it is currently being done, as an early witness-stripped detection may overlook even more ways in which the transaction is invalid (for instance too many sigops).

Crypt-iQ reviewed (commented):
ACK 370770fb361fdb745cfa7bb9abc7ee4a7833b83b modulo commit message change and test comment fix

darosior commented:
Corrected commit message and OP as per @Crypt-iQ's [point](https://github.com/bitcoin/bitcoin/pull/33105#discussion_r2260386539).

darosior commented on src/policy/policy.cpp:
```diff
@@ -337,6 +337,42 @@ bool IsWitnessStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+                continue;
+            }
+            const CScript redeem_script{stack.back().begin(), stack.back().end()};
+            if (redeem_script.IsWitnessProgram(version, program) && !redeem_script.IsPayToAnchor()) {
```
Right. Fixed, thanks.

glozow reviewed (commented):
ACK ffd82f3783d1af4f94f7dbd0af3a80d4591eb185

Crypt-iQ commented:
re-ACK ffd82f3783d1af4f94f7dbd0af3a80d4591eb185

darosior commented on src/test/transaction_tests.cpp:
```diff
@@ -1149,4 +1149,143 @@ BOOST_AUTO_TEST_CASE(max_standard_legacy_sigops)
+
+    // Go through all defined output types and sanity check SpendsNonAnchorWitnessProg.
+
+    // P2PK
```
Sure, done.

Crypt-iQ commented:
re-ACK 27aefac42505e9c083fa131d3d7edbec7803f3c0

glozow reviewed (commented):
reACK 27aefac42505e9c083fa131d3d7edbec7803f3c0
